Implement #224

z33ky opened the pull request:
I'm thinking maybe `BroadPhaseProximityHandler::handle` should be split up into `start` and `stop`, removing the boolean parameter.

z33ky commented:
@sebcrozet Any thoughts?

sebcrozet commented on src/pipeline/broad_phase/broad_phase.rs:
```diff
@@ -41,12 +69,20 @@ pub trait BroadPhase<N: Real, BV, T>: Any + Sync + Send {
     fn deferred_recompute_all_proximities(&mut self);
 
+    /// Updates the object additions, removals, and interferences detection.
+    fn update_with_handler(&mut self, handler: &mut BroadPhaseProximityHandler<T>);
```
This should just replace the existing `update` method to avoid having several ways of doing the same thing.

sebcrozet commented on src/pipeline/broad_phase/broad_phase.rs:
```diff
@@ -24,6 +24,34 @@ impl ProxyHandle {
+    fn handle(&mut self, &T, &T, bool);
+}
+
+struct SimpleBroadPhaseProximityHandler<'a, T: 'a> {
```
This can be removed as this is unnecessary if we don't keep the old `BroadPhase::update` method.

sebcrozet commented on src/pipeline/broad_phase/broad_phase.rs:
```diff
@@ -24,6 +24,34 @@ impl ProxyHandle {
+
+impl<'a, T> BroadPhaseProximityHandler<T> for SimpleBroadPhaseProximityHandler<'a, T> {
+    #[inline]
+    fn allow(&mut self, a: &T, b: &T) -> bool {
```
A more explicit name would be better. For example: `.is_interference_allowed(...)`

sebcrozet commented on src/pipeline/broad_phase/broad_phase.rs:
```diff
@@ -24,6 +24,34 @@ impl ProxyHandle {
+    }
+
+    #[inline]
+    fn handle(&mut self, a: &T, b: &T, started: bool) {
```
I agree splitting this into two methods `.interference_started(...)` and `.interference_stopped(...)` would be better now that we have a trait-object.

sebcrozet commented on src/pipeline/broad_phase/broad_phase.rs:
```diff
@@ -24,6 +24,34 @@ impl ProxyHandle {
 }
 
+/// Proximity handling for BroadPhase updates.
+pub trait BroadPhaseProximityHandler<T> {
```
I think we should not use the `Proximity` word as it may end up being confusing since the concept of `Proximity` already exists on ncollide (for configurations where two solids are closer than a given distance). We can name this `BroadPhaseInterferenceHandler` instead.

z33ky commented:
I also split up `NarrowPhase::handle_interaction` like `BroadPhaseInteferenceHandler::handle_interference`. Seemed like the symmetry makes sense.
I noticed that `DefaultNarrowPhase::interaction_started` does not require `contact_events` or `proximity_events`, so I have also removed the parameters from the `NarrowPhase` trait, though I do not know if this is correct for the general case.

sebcrozet commented:
Thanks! Everything seems good. The split up of `NarrowPhase` is a good call, and removing the contact and proximity events from `interaction_started` do make sense since broad phase are supposed to actually compute proximities and contact on their `update` methods.

I've fixed a conflict with master so I'll have to wait for the CI to pass before merging this.

This is be breaking change.
